File: EmptyProject/WorkflowsTrainingTwo/WorkflowsTrainingCPlus/Parsing/Parser.cpp
#include "Parser.h"
#include "Statements/IfThenElse/IfThenElseNode.h"

using namespace WorkflowsTraining::Parsing;
using namespace WorkflowsTraining::Lexing;
using namespace WorkflowsTraining::Parsing::Statements;

Parser::Parser(std::ifstream& sourceCodeStream)
    : _stream(Lexer::Tokenize(sourceCodeStream)) {
}

ProgramNode Parser::Parse() {
    std::vector<std::unique_ptr<IStatementNode>> statements;

    while (!_stream.EndOfFile()) {
        statements.push_back(Statements::StatementParser::ParseStatement(_stream));
    }

    return ProgramNode(std::move(statements));
}

File: WorkflowResults/WorkflowResults/Parsing/Parser.cs
using WorkflowResults.Lexing;
using WorkflowResults.Parsing.Statements;
using WorkflowResults.Parsing.Statements.Interfaces;

namespace WorkflowResults.Parsing;

public class Parser(StreamReader sourceCodeStream)
{
    private readonly TokenStream _stream = Lexer.Tokenize(sourceCodeStream);

    public ProgramNode Parse()
    {
        IList<IStatementNode> statements = [];
        while (!_stream.EndOfFile())
        {
            statements.Add(StatementParser.ParseStatement(_stream));
        }

        return new ProgramNode(statements);
    }
}